### ISSUE TITLE
Support for custom event handler

### ISSUE THREAD

**Issue by TerenceNg03** [nushell/reedline#520]
## Support for custom event handler

I am currently using `Reedline` to develop a console for a custom language. And I am trying to implement the following feature:
- Automatically add indent when `enter` is hit but validator says input is incomplete.
```
>>> if 3:
...     a = 5
...     a<-- next line should start here
```
- Automatic delete 4 spaces when `delete` is hit. My console treats tab as 4 spaces so I bind `Tab` key to automatically insert 4 spaces but I can not let `delete` to behave correspondingly. For example,  delete 4 space if there are 4 before the cursor, otherwise delete a single character if there is any.
```
>>>     <-- delete all 4 previous space here
>>> asbds<-- only delete `s` from buffer
```

There are also an option that a special event is made for indent handling. But it could be more useful and extensive if user defined handler for `ReedlineEvent` can be supported.

**Comment by sholderbach** [comment]
Yeah having that a bit more systematic than our current escape hatch `ExecuteHostCommand` would be great.

Some of it may be hacked together as it is current practice in nushell thanks to additional ways to write into `Reedline` (https://github.com/nushell/reedline/pull/472)

For examples of the crimes we are pulling in nushell: https://github.com/nushell/nushell/pull/6492

**Comment by actuday6418** [comment]
Any updates on this? Shouldn't we be able to add something simple like an optional callback that gets a reference to the buffer and returns `EditCommand`s? (idk just off the top of my head)
